wire-ios-request-strategy/Sources/Request Strategies/Availability/AvailabilityRequestStrategy.swift

```diff
@@ -18,147 +18,46 @@
+    private let messageSender: MessageSenderInterface
 
-    override public init(withManagedObjectContext managedObjectContext: NSManagedObjectContext, applicationStatus: ApplicationStatus) {
+    public init(context: NSManagedObjectContext, messageSender: MessageSenderInterface) {
+        self.context = context
+        self.modifiedKeysSync = ModifiedKeyObjectSync(trackedKey: AvailabilityKey)
+        self.messageSender = messageSender
 
-        super.init(withManagedObjectContext: managedObjectContext, applicationStatus: applicationStatus)
+        super.init()
 
-        self.modifiedSync = ZMUpstreamModifiedObjectSync(transcoder: self,
-                                                         entityName: ZMUser.entityName(),
-                                                         update: nil,
-                                                         filter: ZMUser.predicateForSelfUser(),
-                                                         keysToSync: [AvailabilityKey],
-                                                         managedObjectContext: managedObjectContext)
+        self.modifiedKeysSync.transcoder = self
     }
 
-    public override func nextRequestIfAllowed(for apiVersion: APIVersion) -> ZMTransportRequest? {
-        return modifiedSync.nextRequest(for: apiVersion)
+    public var contextChangeTrackers: [ZMContextChangeTracker] {
+        return [modifiedKeysSync]
     }
-
-    public var expirationReasonCode: NSNumber?
-
 }
 
-extension AvailabilityRequestStrategy: ZMUpstreamTranscoder {
+extension AvailabilityRequestStrategy: ModifiedKeyObjectSyncTranscoder {
+    typealias Object = ZMUser
 
-    public func request(forUpdating managedObject: ZMManagedObject, forKeys keys: Set<String>, apiVersion: APIVersion) -> ZMUpstreamRequest? {
-        guard let selfUser = managedObject as? ZMUser else { return nil }
+    func synchronize(key: String, for object: ZMUser, completion: @escaping () -> Void) {
+        guard object.isSelfUser else { return completion() }
 
-        let originalPath = "/broadcast/otr/messages"
-        let message = GenericMessage(content: WireProtos.Availability(selfUser.availability))
+        let message = GenericMessage(content: WireProtos.Availability(object.availability))
         let recipients = ZMUser.recipientsForAvailabilityStatusBroadcast(in: context, maxCount: maximumBroadcastRecipients)
+        let proteusMessage = GenericMessageEntity(message: message, context: context, targetRecipients: .users(recipients), completionHandler: nil)
 
-        guard let dataAndMissingClientStrategy = message.encryptForTransport(forBroadcastRecipients: recipients, in: context) else {
-            return nil
+        WaitingGroupTask(context: context) { [self] in
+            try? await messageSender.broadcastMessage(message: proteusMessage)
+            await context.perform { [self] in
+                completion()
+                context.enqueueDelayedSave()
```

Comment:
normal to call enqueue after completion?

Reply:
Inside the completion the `modifiedKeys` of the user are reset we need to save the context after that. Normally this save happen implicitly after http response processing id complete, but since this is happening async we need to manually trigger a save.

Reply:
Could you put this explanation into a comment at this place, please?
I also tend to make any completion call the last statement in the block when I encounter something like this.

Reply:
Done